include/deal.II/non_matching/fe_immersed_values.h

```diff
@@ -0,0 +1,145 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2021 - 2021 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#ifndef dealii_non_matching_fe_immersed_values_h
+#define dealii_non_matching_fe_immersed_values_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/dofs/dof_accessor.h>
+
+#include <deal.II/fe/fe.h>
+#include <deal.II/fe/fe_update_flags.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/mapping.h>
+
+#include <deal.II/grid/tria_iterator.h>
+
+#include <deal.II/non_matching/immersed_surface_quadrature.h>
+
+DEAL_II_NAMESPACE_OPEN
+namespace NonMatching
+{
+  /**
+   * Finite element evaluated in the quadrature points of an
+   * ImmersedSurfaceQuadrature of a cell.
+   *
+   * The shape functions values and their derivatives are the same as for an
+   * FEValues-object, but the JxW-values are computed with the transformation
+   * described in the documentation of ImmersedSurfaceQuadrature. Further, the
+   * normal_vector-function returns the normal to the immersed surface.
+   *
+   * The reinit-function of this class exist mostly to be consistent with the
+   * other FEValues-like classes. The immersed quadrature rule will typically
+   * vary between each cell of the triangulation. Thus, an
+   * FEImmersedSurfaceValues object can, typically, not be reused for different
+   * cells.
+   *
+   * See also documentation in FEValuesBase.
+   *
+   * @ingroup feaccess
+   */
+  template <int dim>
+  class FEImmersedSurfaceValues : public FEValuesBase<dim, dim>
+  {
+  public:
+    /**
+     * Constructor. Gets cell-independent data from mapping and finite element
+     * objects, matching the quadrature rule and update flags.
+     *
+     * @note Currently this class is only implemented for MappingCartesian.
+     */
+    FEImmersedSurfaceValues(const Mapping<dim> &                  mapping,
+                            const FiniteElement<dim> &            element,
+                            const ImmersedSurfaceQuadrature<dim> &quadrature,
+                            const UpdateFlags                     update_flags);
+
+    /**
+     * Reinitialize the gradients, Jacobi determinants, etc for the incoming @p cell.
+     */
+    template <bool level_dof_access>
+    void
+    reinit(
+      const TriaIterator<DoFCellAccessor<dim, dim, level_dof_access>> &cell);
+
```

Comment:
Do you anticipate that we never need an initialization with `typename Triangulation<dim, spacedim>::cell_iterator &cell`, as we provide in the `FEValues` class besides the `TriaIterator<DoFCellAccessor>`?

Reply:
I've added another `reinit` function and modified the test slightly to check that it works.